src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs

```diff
@@ -230,6 +211,187 @@ public override void EndWrite(IAsyncResult asyncResult)
+        private protected sealed class SyncAsyncWorkItem : IThreadPoolWorkItem, ICriticalNotifyCompletion
+        {
+            /// <summary>A thread handle for the current OS thread.</summary>
+            /// <remarks>This is lazily-initialized for the current OS thread. We rely on finalization to clean up after it when the thread goes away.</remarks>
+            [ThreadStatic]
+            private static SafeThreadHandle? t_currentThreadHandle;
+
+            /// <summary>The OS handle of the thread performing the I/O.</summary>
+            public SafeThreadHandle? ThreadHandle;
+
+            /// <summary>Whether the call to CancellationToken.UnsafeRegister completed.</summary>
+            public volatile bool FinishedCancellationRegistration;
+            /// <summary>Whether the I/O operation has finished (successfully or unsuccessfully) and is requesting cancellation attempts stop.</summary>
+            public volatile bool ContinueTryingToCancel = true;
+            /// <summary>The Action continuation object handed to this instance when used as an awaiter to scheduler work to the thread pool.</summary>
+            private Action? _continuation;
+
+            // awaitable / awaiter implementation that enables this instance to be awaited in order to queue
+            // execution to the thread pool.  This is purely a cost-saving measure in order to reuse this
+            // object we already need as the queued work item.
+            public SyncAsyncWorkItem GetAwaiter() => this;
+            public bool IsCompleted => false;
+            public void GetResult() { }
+            public void OnCompleted(Action continuation) => throw new NotSupportedException();
+            public void UnsafeOnCompleted(Action continuation)
+            {
+                Debug.Assert(_continuation is null);
+                _continuation = continuation;
+                ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: true);
+            }
+            void IThreadPoolWorkItem.Execute() => _continuation!();
+
+            /// <summary>Registers for cancellation with the specified token.</summary>
+            /// <remarks>Upon cancellation being requested, the implementation will attempt to CancelSynchronousIo for the thread calling RegisterCancellation.</remarks>
+            public CancellationTokenRegistration RegisterCancellation(CancellationToken cancellationToken)
+            {
+                // If the token can't be canceled, there's nothing to register.
+                if (!cancellationToken.CanBeCanceled)
+                {
+                    return default;
+                }
+
+                // Get a handle for the current thread. This is stored and used to cancel the I/O on this thread
+                // in response to the cancellation token having cancellation requested.  If the handle is invalid,
+                // which could happen if OpenThread fails, skip attempts at cancellation.
```

Comment:
nit: you could mention that `THREAD_TERMINATE` access right is required by the `CancelSynchronousIo` sys-call